_posts/2016-06-19-graphql-for-mobile.markdown

```diff
@@ -0,0 +1,99 @@
+
+This blog post covers our usage of GraphQL, and what I've learned in the last 3 months of using it in [Eigen](https://github.com/artsy/eigen/).
+
+<!-- more -->
+
+### So what is GraphQL
+
+So, you can get the full explaination on [their website](http://graphql.org). Though I found running through [Learn GraphQL](https://learngraphql.com) to really hammer down how it works as an API consumer. Reading the [introduction blog post](https://facebook.github.io/react/blog/2015/05/01/graphql-introduction.html) can be useful too, I'll be doing a quick TLDR though.
+
+GraphQL is an API middle-layer. It acts as an intermediate layer between multiple front-end clents and multiple back-end APIs. This means it can easily coalesce multiple API calls into a single request, this can be a _massive_ user experience improvement when you have a screen that requires information from multiple sources before you can present anything to a user.
+
+<img src="/images/2016-06-19-graphql-for-iOS-devs/graphQL.svg" width=100%>
+
+As a client, you [send](https://github.com/artsy/eigen/blob/dac7c80b66b600f9a45aaae6095544fe420f0bbc/Artsy/Networking/ARRouter.m#L1011) a "JSON-SQL"-like query structure, which is heirarchical and easy to read:
+
+```json
+{
+  artwork(id: "kimber-berry-as-close-to-magic-as-you-can-get") {
+    id
+    additional_information
+
+    is_price_hidden
+    is_inquireable
+  }
+}
+
+```
+
+> This will search for a [specific artwork](https://www.artsy.net/artwork/kimber-berry-as-close-to-magic-as-you-can-get), sending back the Artwork's `id`, `additional_information`, `is_price_hidden` and `is_inquireable`.
+
+It's important to note here, the data being sent _back_ is only what you ask for. This is not defined on the server as a _short_ or _embedded_ version of a model, but the specific data the client requested. When bandwidth and speed is crucial, this is the other way in which GraphQL improves the end-user experience.
+
+So, that's the two killer features:
+
+1. Coelesce Multiple Network Requests.
+2. Only Send The Data You Want.
+
+This is in stark contrast to existing API concepts, like [HAL](http://stateless.co/hal_specification.html) and [JSON-API](http://jsonapi.org) - both of which are optimised for caching, and rely on "one model, one request" types of API access. E.g. a list of Artworks would actually contain a list of hrefs instead of the model data, and you have to fetch each model as a separate request.
+
+### Using GraphQL
+
+Artsy's GraphQL server is (unsurprisingly) open-source, it's at [artsy/metaphysics](https://github.com/artsy/metaphysics). However, it's not publicly accessible, ([yet?](https://github.com/artsy/metaphysics/issues/279)). One of the coolest things about developing against a GraphQL server is that it has a built-in IDE. I can't show you ours, but I can send you to [Clay Allsop's](http://clayallsopp.com) [GraphQLHub](https://www.graphqlhub.com):
```

Comment:
GraphiQL is not built into GraphQL. GraphQL is just a concept, metaphysics is an implementation thereof, GraphiQL is a popular IDE to use with a GraphQL server.

The more important thing here is the degree of reflection (on the schema and types) that the GraphQL spec prescribes, which makes awesome IDEs like this feasible at all.